File: 22 Generate Parentheses/generate_parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        def backtrack(res, open_num = 0, close_num = 0, curr= ""):
            if len(curr) == n * 2:
                res.append(curr)
            else:
                if open_num < n:
                    backtrack(res, open_num + 1, close_num, curr + "(")
                if close_num < open_num:
                    backtrack(res, open_num, close_num + 1, curr + ")")
        
        backtrack(res)
        return res

File: 0022 Generate Parentheses/GenerateParentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> res = new ArrayList<>();
        backtrack(res, new StringBuilder(), 0, 0, n);
        return res;
    }
    
    private void backtrack(List<String> res, StringBuilder curr, int left, int right, int n) {
        if (curr.length() == n * 2) {
            res.add(curr.toString());
        }
        if (left < n) {
            curr.append("(");
            backtrack(res, curr, left + 1, right, n);
            curr.deleteCharAt(curr.length() - 1);
        }
        if (right < left) {
            curr.append(")");
            backtrack(res, curr, left, right + 1, n);
            curr.deleteCharAt(curr.length() - 1);
        }       
    }
}